src/dali_executor/dali_executor.cc

```diff
@@ -55,29 +55,59 @@ IDescr DaliExecutor::ScheduleInputCopy(const IDescr& input) {
   assert(input.buffers.size() > 0);
   IOBufferI* buffer;
   if (input.buffers[0].device == device_type_t::CPU) {
-    buffer = &cpu_buffers_[input.meta.name];
+    buffer = &cpu_buffers_[input.meta.name + "_inp"];
   } else {
-    buffer = &gpu_buffers_[input.meta.name];
+    buffer = &gpu_buffers_[input.meta.name + "_inp"];
   }
   size_t size = 0;
   for (auto& buf : input.buffers)
     size += buf.size;
   buffer->resize(size);
   auto descriptor = buffer->get_descr();
   char* dst = reinterpret_cast<char*>(descriptor.data);
+  auto stream = pipeline_.CopyStream();
   for (auto& buf : input.buffers) {
     thread_pool_.AddWork(
-        [descriptor, dst, buf](int) {
-          MemCopy(descriptor.device, dst, buf.device, buf.data, buf.size);
+        [stream, descriptor, dst, buf](int) {
+          MemCopy(descriptor.device, dst, buf.device, buf.data, buf.size, stream);
         },
         buf.size, true);
     dst += buf.size;
   }
   return IDescr{input.meta, {descriptor}};
 }
 
-void DaliExecutor::RunInputCopy() {
+void DaliExecutor::ScheduleOutputCopy(const ODescr& output, int output_idx) {
+  const auto& name = output.meta.name;
+  const auto& out_buffers = output.buffers;
+  size_t size = 0;
+  for (auto& out_buff : out_buffers) {
+    size += out_buff.size;
+  }
+  IOBufferI* interm_buffer;
+  if (pipeline_.GetOutputDevice(output_idx) == device_type_t::CPU) {
+    interm_buffer = &cpu_buffers_[name + "_out"];
```

Comment:
Same here.

Reply:
Added GetInputBuffer and GetOutputBuffer